lucene/core/src/test/org/apache/lucene/util/TestBitUtil.java

```diff
@@ -0,0 +1,79 @@
+ * limitations under the License.
+ */
+
+package org.apache.lucene.util;
+
+public class TestBitUtil extends LuceneTestCase {
+
+  public void testNextBitSet() {
+    for (int i = 0; i < 10000; i++) {
+      long[] bits = buildRandomBits();
+      int numLong = bits.length - 1;
+
+      // Verify nextBitSet with countBitsUpTo for all bit indexes.
+      for (int bitIndex = -1; bitIndex < 64 * numLong; bitIndex++) {
+        int nextIndex = BitUtil.nextBitSet(bits, numLong, bitIndex);
+        if (nextIndex == -1) {
+          assertEquals("No next bit set, so expected no bit count diff (i=" + i + " bitIndex=" + bitIndex + ")",
+              BitUtil.countBitsUpTo(bits, numLong, bitIndex + 1), BitUtil.countBits(bits, numLong));
+        } else {
+          assertTrue("Expected next bit set at nextIndex=" + nextIndex + " (i=" + i + " bitIndex=" + bitIndex + ")",
+              BitUtil.isBitSet(bits, numLong, nextIndex));
+          assertEquals("Next bit set at nextIndex=" + nextIndex + " so expected bit count diff of 1 (i=" + i + " bitIndex=" + bitIndex + ")",
+              BitUtil.countBitsUpTo(bits, numLong, bitIndex + 1) + 1, BitUtil.countBitsUpTo(bits, numLong, nextIndex + 1));
+        }
+      }
+    }
+  }
+
+  public void testPreviousBitSet() {
+    for (int i = 0; i < 10000; i++) {
+      long[] bits = buildRandomBits();
+      int numLong = bits.length - 1;
+
+      // Verify previousBitSet with countBitsUpTo for all bit indexes.
+      for (int bitIndex = 0; bitIndex <= 64 * numLong; bitIndex++) {
+        int previousIndex = BitUtil.previousBitSet(bits, numLong, bitIndex);
+        if (previousIndex == -1) {
+          assertEquals("No previous bit set, so expected bit count 0"
+                  + " (i=" + i + " bitIndex=" + bitIndex + ")",
+              0, BitUtil.countBitsUpTo(bits, numLong, bitIndex));
+        } else {
+          assertTrue("Expected previous bit set at previousIndex=" + previousIndex
+                  + " (i=" + i + " bitIndex=" + bitIndex + ")",
+              BitUtil.isBitSet(bits, numLong, previousIndex));
+          int bitCount = BitUtil.countBitsUpTo(bits, numLong, Math.min(bitIndex + 1, numLong * Long.SIZE));
+          int expectedPreviousBitCount = bitIndex < numLong * Long.SIZE && BitUtil.isBitSet(bits, numLong, bitIndex) ?
+              bitCount - 1 : bitCount;
+          assertEquals("Previous bit set at previousIndex=" + previousIndex
+                  + " with current bitCount=" + bitCount
+                  + " so expected previousBitCount=" + expectedPreviousBitCount
+                  + " (i=" + i + " bitIndex=" + bitIndex + ")",
+              expectedPreviousBitCount, BitUtil.countBitsUpTo(bits, numLong, previousIndex + 1));
+        }
+      }
+    }
+  }
+
+  private long[] buildRandomBits() {
+    long[] bits = new long[random().nextInt(3) + 2];
+    for (int j = 0; j < bits.length; j++) {
+      bits[j] = random().nextLong();
```

Comment:
nit: this way of generating longs makes it very unlikely to have zeros, maybe bias it towards zeros with e.g.

```suggestion
      bits[j] = random().nextBoolean() ? 0L : random().nextLong(); // bias towards zeros which require special logic
```

Reply:
Good point.